### PR TITLE
add production and sandbox security

### PR DESCRIPTION
**Pull request by tharindu1st**
Resolves #https://github.com/wso2/product-apim/issues/1547

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/template/SecurityConfigContext.java [hunk 1]
```diff
@@ -16,45 +16,171 @@
...
+                } else {
+                    endpointSecurityModel.setType(APIConstants.ENDPOINT_SECURITY_TYPE_DIGEST);
+                }
+                endpointSecurityModel.setAlias(alias);
+                String unpw = api.getEndpointUTUsername() + ":" + api.getEndpointUTPassword();
+                endpointSecurityModel.setBase64EncodedPassword(new String(Base64.encodeBase64(unpw.getBytes())));
+                endpointSecurityModelMap.put(APIConstants.ENDPOINT_SECURITY_PRODUCTION, endpointSecurityModel);
+                endpointSecurityModelMap.put(APIConstants.ENDPOINT_SECURITY_SANDBOX, endpointSecurityModel);
+            } else {
+                if (api.getEndpointConfig() != null && !"".equals(api.getEndpointConfig())) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/template/SecurityConfigContext.java]: use string utils

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -1307,6 +1299,75 @@ public void updateAPI(API api) throws APIManagementException, FaultGatewaysExcep
         }
     }
 
+    private void updateEndpointSecurity(API oldApi, API api) throws APIManagementException {
+        try {
+            if (api.isEndpointSecured() && StringUtils.isBlank(api.getEndpointUTPassword()) &&
+                    !StringUtils.isBlank(oldApi.getEndpointUTPassword())) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Given endpointsecurity password is empty");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: put some context llike api name

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 3]
```diff
@@ -1307,6 +1299,75 @@ public void updateAPI(API api) throws APIManagementException, FaultGatewaysExcep
...
+                                                .parse(new ObjectMapper().writeValueAsString(endpointSecurity)));
+                            }
+                            endpointConfigJson.replace(APIConstants.ENDPOINT_SECURITY,endpointSecurityJson);
+                        }
+                    }
+                    api.setEndpointConfig(endpointConfigJson.toJSONString());
+                }
+            }
+        } catch (ParseException | JsonProcessingException e) {
+            throw new APIManagementException("Error while processing endpoint security", e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: add more context like api name

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/template/SecurityConfigContext.java [hunk 4]
```diff
@@ -16,45 +16,171 @@
...
+    }
+
+    @Override
+    public void validate() throws APITemplateException, APIManagementException {
+
+        super.validate();
+        if (api != null) {
+            JSONParser parser = new JSONParser();
+            //check if endpoint config exists
+            String config_json = api.getEndpointConfig();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/template/SecurityConfigContext.java]: why not camel case variable types.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/template/SecurityConfigContext.java [hunk 5]
```diff
@@ -16,45 +16,171 @@
...
+
+            if (config_json != null && !"".equals(config_json)) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/template/SecurityConfigContext.java]: string utils

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIProductResource.java [hunk 6]
```diff
@@ -130,4 +133,15 @@ public void setEndpointConfig(String endpointConfig) {
...
         return apiIdentifier.getApiName() + "--v" + apiIdentifier.getVersion();
     }
+
+    public Map<String, EndpointSecurity> getEndpointSecurityMap() {
+
+        return endpointSecurityMap;
+    }
+
+    public void setEndpointSecurityMap(
+            Map<String, EndpointSecurity> endpointSecurityMap) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIProductResource.java]: can bring to the same line i think.

**Pull request merged**